crates/ruff_server/src/server/api/requests/execute_command.rs

```diff
@@ -0,0 +1,144 @@
+use std::collections::HashMap;
+
+use crate::server;
+use crate::server::api::LSPResult;
+use crate::server::client;
+use crate::server::schedule::Task;
+use crate::session::Session;
+use crate::DIAGNOSTIC_NAME;
+use lsp_server::ErrorCode;
+use lsp_types::{self as types, request as req};
+use serde::Deserialize;
+
+enum Command {
+    Format,
+    FixAll,
+    OrganizeImports,
+}
+
+pub(crate) struct ExecuteCommand;
+
+#[derive(Deserialize)]
+struct TextDocumentArgument {
+    uri: types::Url,
+}
+
+impl super::RequestHandler for ExecuteCommand {
+    type RequestType = req::ExecuteCommand;
+}
+
+impl super::SyncRequestHandler for ExecuteCommand {
+    fn run(
+        session: &mut Session,
+        _notifier: client::Notifier,
+        requester: &mut client::Requester,
+        params: types::ExecuteCommandParams,
+    ) -> server::Result<Option<serde_json::Value>> {
+        let Some(command) = Command::from_str(&params.command) else {
+            return Err(anyhow::anyhow!("")).with_failure_code(ErrorCode::InvalidParams);
+        };
+
+        let mut changes = HashMap::new();
+        let documents =
+            args_as_text_documents(params.arguments).with_failure_code(ErrorCode::InvalidParams)?;
+        for document in documents {
+            let snapshot = session
+                .take_snapshot(&document.uri)
```

Comment:
Nit: We could incline `args_as_text_documents` to avoid the vec allocation and without increasing the complexity (I almost find it easier to read because I don't need to jump to `args_as_text_documents` to understand what's happening)
```suggestion
        for argument in params.arguments {
            let document = serde_json::from_value(argument).with_failure_code(ErrorCode::InvalidParams)?;
            let snapshot = session
                .take_snapshot(&document.uri)
```